spec/components/cards_spec.rb

```diff
@@ -0,0 +1,280 @@
+require "rails_helper"
+
+describe "Cards", type: :view do
+  def component_name
+    "cards"
+  end
+
+  it "renders nothing when no data is provided" do
+    assert_empty render_component({})
+  end
+
+  it "renders component wrapper" do
+    test_data = {
+      items: [
+        {
+          link: {
+            text: "Benefits",
+            path: "http://www.gov.uk",
+          },
+        },
+      ],
+    }
+    render_component(test_data)
+    assert_select ".gem-c-cards", count: 1
+  end
+
+  it "renders list heading" do
+    test_data = {
+      heading: "Topics",
+      items: [
+        {
+          link: {
+            text: "Benefits",
+            path: "http://www.gov.uk",
+          },
+        },
+      ],
+    }
+    render_component(test_data)
+    assert_select ".gem-c-cards__heading.govuk-heading-m", text: "Topics"
+  end
+
+  it "renders list heading using default heading level" do
+    test_data = {
+      heading: "Topics",
+      items: [
+        {
+          link: {
+            text: "Benefits",
+            path: "http://www.gov.uk",
+          },
+        },
+      ],
+    }
+    render_component(test_data)
+    assert_select "h2.gem-c-cards__heading.govuk-heading-m", count: 1
+  end
+
+  it "renders list heading using custom heading level" do
+    test_data = {
+      heading: "Topics",
+      heading_level: 3,
+      items: [
+        {
+          link: {
+            text: "Benefits",
+            path: "http://www.gov.uk",
+          },
+        },
+      ],
+    }
+    render_component(test_data)
+    assert_select "h3.gem-c-cards__heading.govuk-heading-m", count: 1
+  end
+
+  it "renders list" do
```

Comment:
I think again that this wouldn't follow the single outcome principle for unit tests (see https://github.com/alphagov/govuk_publishing_components/pull/2624#discussion_r811979265).

Reply:
I don't think the tests need to be that granular. You could do a single test for 'all expected elements are present' and it would be clear enough that the one test was checking for multiple things, and it would be obvious where it had failed if it failed. Test files tend to get pretty long so I think making them smaller if possible is a good idea (I've had to completely rewrite some long test files and it can be hard going).

Something you might not be aware of - there's a [general test](https://github.com/alphagov/govuk_publishing_components/blob/master/spec/components/all_components_spec.rb#L26) that checks to see that every component template contains a class of the form `gem-c-[something]`, so you don't need to explicitly write a test to check for that.

Reply:
See https://github.com/alphagov/govuk_publishing_components/pull/2624#discussion_r816938825